lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+                    flat_vec.append(&mut flat);
+                } else {
+                    for elem in tbl.sequence_values::<LuaValue>() {
+                        flat_vec.push(elem?);
+                    }
+                }
+            }
+            LuaValue::Nil => (),
+            other => {
+                flat_vec.push(other);
+            }
+        }
+    }
+    Ok(flat_vec)
+}
+
+/// note that the `#` operator only works correctly on arrays in Lua
+fn count<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<usize> {
+    Ok(table.pairs::<LuaValue, LuaValue>().count())
+}
+
+fn get<'lua>(
+    _: &'lua Lua,
+    (table, key, mut extra_keys): (Table<'lua>, LuaValue<'lua>, LuaMultiValue<'lua>),
+) -> mlua::Result<LuaValue<'lua>> {
+    if extra_keys.is_empty() {
+        return Ok(table.get::<_, LuaValue>(key)?);
+    }
+
+    let mut value: LuaValue = table.get(key.clone())?;
+
+    let mut value_tbl = match table.get::<_, Table>(key) {
+        Ok(t) => t,
+        Err(_) => return Ok(LuaValue::Nil), // if extra_keys were empty, we wouldn't get here
+    };
+
+    while let Some(next_key) = extra_keys.pop_front() {
+        value = value_tbl.get(next_key.clone())?;
+        let new_val_tbl = value_tbl.get::<_, Table>(next_key);
+        value_tbl = match new_val_tbl {
+            Ok(t) => t,
+            Err(_) => {
+                if extra_keys.is_empty() {
+                    return Ok(value);
+                } else {
+                    return Ok(LuaValue::Nil);
+                }
+            }
+        }
+    }
+
+    Ok(value)
+}
+
+fn has_key<'lua>(
+    _: &'lua Lua,
+    (table, key, mut extra_keys): (Table<'lua>, LuaValue, LuaMultiValue),
+) -> mlua::Result<bool> {
+    if extra_keys.is_empty() {
+        return Ok(table.contains_key(key)?);
+    }
+
+    let mut value_has_key = table.contains_key(key.clone())?;
+
+    let mut value = match table.get::<_, Table>(key) {
+        Ok(t) => t,
+        Err(_) => return Ok(false), // if extra_keys were empty, we wouldn't get here
+    };
+
+    while let Some(next_key) = extra_keys.pop_front() {
+        value_has_key = value.contains_key(next_key.clone())?;
+        let new_val = value.get::<_, Table>(next_key);
+        value = match new_val {
+            Ok(t) => t,
+            Err(_) => return Ok(value_has_key && extra_keys.is_empty()),
+        };
+    }
+
+    Ok(value_has_key)
+}
+
+fn has_value<'lua>(
+    lua: &'lua Lua,
+    (table, value, behavior): (Table<'lua>, LuaValue, Option<DepthMode>),
+) -> mlua::Result<bool> {
+    let behavior = behavior.unwrap_or_default();
+    match behavior {
+        DepthMode::Top => {
+            // we don't need a clone in this case
+            for pair in table.pairs::<LuaValue, LuaValue>() {
+                let (_, tbl_value) = pair?;
+                if tbl_value == value {
+                    return Ok(true);
+                }
+            }
+        }
+        DepthMode::Deep => {
```

Comment:
I can just remove it to simplify the code for now. I thought it might be useful to see if certain values were precent in one's config, but in that case you would usually know what key it should be set for anyways.

Reply:
Actually looking at the updated code, I don't really think it adds much extra. Though I do agree that it will probably only be rarely used. Do you prefer to just remove it?